src/python/pants/backend/python/target_types_rules.py

```diff
@@ -135,29 +142,124 @@ async def inject_pex_binary_entry_point_dependency(
+                    )
+
+                for target in targets:
+                    if target.has_field(PexEntryPointField):
+                        binary_entry_point = await Get(
+                            ResolvedPexEntryPoint,
+                            ResolvePexEntryPointRequest(target[PexEntryPointField]),
+                        )
+                        ep = binary_entry_point.val
+                        break
+                else:
+                    raise InvalidFieldException(
+                        f'The "entry_points" target address {repr(ref)} does not resolve to a pex_binary() target, but:'
+                        + ", ".join(str(t) for t in targets)
+                    )
+            else:
+                ep = EntryPoint.parse(ref, f"{name} for {address} {section}")
+
+            if ep is not None:
+                entry_points[section][name] = ep
+
+    return ResolvedPythonDistributionEntryPoints(
+        FrozenDict({key: FrozenDict(value) for key, value in entry_points.items()})
+    )
+
+
 class InjectPythonDistributionDependencies(InjectDependenciesRequest):
     inject_for = PythonDistributionDependencies
 
 
 @rule
 async def inject_python_distribution_dependencies(
-    request: InjectPythonDistributionDependencies,
+    request: InjectPythonDistributionDependencies, python_infer_subsystem: PythonInferSubsystem
 ) -> InjectedDependencies:
-    """Inject any `.with_binaries()` values, as it would be redundant to have to include in the
-    `dependencies` field."""
+    """Inject dependencies that we can infer from entry points in the distribution.
+
+    Inject module owners referred to from `entry_points`.
+
+    Inject any `.with_binaries()` values, as it would be redundant to have to
+    include in the `dependencies` field.
+    """
     original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
-    with_binaries = original_tgt.target[PythonProvidesField].value.binaries
-    if not with_binaries:
-        return InjectedDependencies()
-    # Note that we don't validate that these are all `pex_binary` targets; we don't care about
-    # that here. `setup_py.py` will do that validation.
-    addresses = await Get(
-        Addresses,
-        UnparsedAddressInputs(
-            with_binaries.values(), owning_address=request.dependencies_field.address
+
+    entry_points = await Get(
+        ResolvedPythonDistributionEntryPoints,
+        ResolvePythonDistributionEntryPointsRequest(
+            original_tgt.target[PythonDistributionEntryPoints]
         ),
     )
-    return InjectedDependencies(addresses)
+
+    entry_point_addresses = Addresses()
+
+    if entry_points.val and python_infer_subsystem.entry_points:
+        address = original_tgt.target.address
+        explicitly_provided_deps = await Get(
+            ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])
+        )
```

Comment:
It'd be good to do this sooner via a heterogeneous `MultiGet` on the above `Get`: https://www.pantsbuild.org/docs/rules-api-tips#tip-use-multiget-for-increased-concurrency.